apm-agent-plugins/apm-javalin-plugin/src/main/java/co/elastic/apm/agent/javalin/JavalinInstrumentation.java

```diff
@@ -0,0 +1,167 @@
+import static net.bytebuddy.matcher.ElementMatchers.isInterface;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.not;
+import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
+
+public class JavalinInstrumentation extends TracerAwareInstrumentation {
+
+    private static final String FRAMEWORK_NAME = "Javalin";
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return hasSuperType(named("io.javalin.http.Handler")).and(not(isInterface()));
+    }
+
+    @Override
+    public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+        return named("handle").and(takesArgument(0, named("io.javalin.http.Context")));
+    }
+
+    @Override
+    public Collection<String> getInstrumentationGroupNames() {
+        return Collections.singleton("javalin");
+    }
+
+    @Override
+    public String getAdviceClassName() {
+        return "co.elastic.apm.agent.javalin.JavalinInstrumentation$HandlerAdapterAdvice";
+    }
+
+    public static class HandlerAdapterAdvice {
+
+        private static Field handlerTypeField = null;
+
+        @Advice.OnMethodEnter(suppress = Throwable.class, inline = false)
+        public static Object setSpanAndTransactionName(@Advice.This Handler handler,
+                                                       @Advice.Argument(0) Object ctxObject) throws Exception {
+            final Transaction transaction = tracer.currentTransaction();
+            if (transaction == null) {
+                return null;
+            }
+            Context ctx = ((Context) ctxObject);
+
+            // TODO FIXME I don't think this is the right way, I am sure there is a fancy bytecode trick that I am not aware off
+            if (handlerTypeField == null) {
+                Field field = ctx.getClass().getDeclaredField("handlerType");
+                field.setAccessible(true);
+                handlerTypeField = field;
+            }
+
+            final HandlerType handlerType = (HandlerType) handlerTypeField.get(ctx);
+            final String handlerClassName = handler.getClass().getName();
+
+            // transaction name like the HTTP path
+            if (handlerType.isHttpMethod()) {
+                transaction.setFrameworkName(FRAMEWORK_NAME);
+                transaction.setFrameworkVersion(VersionUtils.getVersion(Handler.class, "io.javalin", "javalin"));
+                // no need for anonymous handler class names in the transaction
+                if (handlerClassName.startsWith("co.elastic.apm.agent.javalin.JavalinHandlerLambdaInstrumentation$WrappingHandler")
+                    || handlerClassName.startsWith("io.javalin.http.JavalinServlet")) {
+                    setName(transaction, handlerType.name() + " " + ctx.endpointHandlerPath());
+                } else {
+                    setName(transaction, handlerType.name() + " " + ctx.endpointHandlerPath() + " " + handlerClassName);
```

Comment:
Don't use these `+` operators for names, as this allocates unnecessary objects.
This misses the purpose of our `co.elastic.apm.agent.impl.transaction.AbstractSpan#appendToName()` API. Your `setName()` method can get all required name elements and build it through appends.